x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/common/EmbeddingRequestChunker.java

```diff
@@ -160,33 +187,82 @@ private class DebatchingListener implements ActionListener<InferenceServiceResul
+                case BYTE -> handleByteResults(inferenceServiceResults);
+            }
+            ;
+        }
+
+        private void handleFloatResults(InferenceServiceResults inferenceServiceResults) {
+            if (inferenceServiceResults instanceof InferenceTextEmbeddingFloatResults floatEmbeddings) {
+                if (failIfNumRequestsDoNotMatch(floatEmbeddings.embeddings().size())) {
                     return;
                 }
 
                 int start = 0;
                 for (var pos : positions) {
-                    results.get(pos.inputIndex())
-                        .setOnce(pos.chunkIndex(), textEmbeddingResults.embeddings().subList(start, start + pos.embeddingCount()));
+                    floatResults.get(pos.inputIndex())
+                        .setOnce(pos.chunkIndex(), floatEmbeddings.embeddings().subList(start, start + pos.embeddingCount()));
                     start += pos.embeddingCount();
                 }
+
+                if (resultCount.incrementAndGet() == totalNumberOfRequests) {
+                    sendResponse();
+                }
+            } else {
+                onFailure(
+                    unexpectedResultTypeException(inferenceServiceResults.getWriteableName(), InferenceTextEmbeddingFloatResults.NAME)
+                );
             }
+        }
 
-            if (resultCount.incrementAndGet() == totalNumberOfRequests) {
-                sendResponse();
+        private void handleByteResults(InferenceServiceResults inferenceServiceResults) {
+            if (inferenceServiceResults instanceof InferenceTextEmbeddingByteResults byteEmbeddings) {
+                if (failIfNumRequestsDoNotMatch(byteEmbeddings.embeddings().size())) {
+                    return;
+                }
+
+                int start = 0;
+                for (var pos : positions) {
+                    byteResults.get(pos.inputIndex())
+                        .setOnce(pos.chunkIndex(), byteEmbeddings.embeddings().subList(start, start + pos.embeddingCount()));
+                    start += pos.embeddingCount();
+                }
+
+                if (resultCount.incrementAndGet() == totalNumberOfRequests) {
+                    sendResponse();
+                }
+            } else {
+                onFailure(
+                    unexpectedResultTypeException(inferenceServiceResults.getWriteableName(), InferenceTextEmbeddingByteResults.NAME)
+                );
             }
         }
 
+        private boolean failIfNumRequestsDoNotMatch(int numberOfResults) {
+            int numberOfRequests = positions.stream().mapToInt(SubBatchPositionsAndCount::embeddingCount).sum();
+            if (numberOfRequests != numberOfResults) {
+                onFailure(
+                    new ElasticsearchStatusException(
+                        "Error the number of embedding responses [{}] does not equal the number of " + "requests [{}]",
+                        RestStatus.BAD_REQUEST,
```

Comment:
Is this a bad request? I can't really imagine how this would happen, but would it be because of the request?

Reply:
I changed to internal error as it should not happen